### PR TITLE
Support container insights for AOC

### PR DESCRIPTION
**Pull request by pxaws**
**Description:** 
Support aws eks container insights for AOC

This commit migrates the code from cloudwatch agent to support container insights for AOC. Three submodules are introduced:
* k8scommon
  * utils functions to translate metric to OT metrics
  * constants for container insight use case
  * utils functions to call k8s api server
  * utils functions to call kubelet apis
* awscontainerinsightreceiver
  * use k8sapiserver to get cluster-level metrics (using leader election to make sure only one node is generating the metrics)
  * use cadvisor to get various cpu/mem/filesystem/network metrics and decorate them with relevant pod information (so that awscontainerinsightprocessor can process them) 
  * collect host information (like ebs volume, instance id/type) to decorate the metrics
* awscontainerinsightprocessor
  * service store retrieves k8s service information and is used to add service info to metrics
  * pod store retrieves pod metadata and is used to decorate various pod/container level metrics and to produce some new metrics

The final goal is to contribute those components to upstream repo: https://github.com/open-telemetry/opentelemetry-collector-contrib. However this will take a long time considering the amount of code changes. So the current plan is to merge it in AOC first and then do the release.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** < Describe what testing was performed and which tests were added.>

**Documentation:** < Describe the documentation added.>

### CODE REVIEW

File: internal/aws/k8scommon/kubeletutil/kubeletclient.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (wyTrivail) [inline internal/aws/k8scommon/kubeletutil/kubeletclient.go:1]: license header need to be changed

**Review by wyTrivail — COMMENTED** [review]
seems we are copying the code from cwagent, i wonder should we keep the license headers. any memory consumption change since we add too many files?

File: internal/aws/k8scommon/kubeletutil/kubeletclient.go (hunk 1 above)
Comment (pxaws) [reply internal/aws/k8scommon/kubeletutil/kubeletclient.go:1]: Sure. This is a work-in-progress PR. I will add more changes

Comment (pxaws) [reply internal/aws/k8scommon/kubeletutil/kubeletclient.go:1]: The added files are only for container insight use case. So it won't affect existing memory consumption because those components will be not enabled. The final goal is to upstream those changes to opentelemetry-contrib. But it might takes some time, so for now, we plan to vendor the code in AOC temporarily.

File: internal/aws/k8scommon/k8sclient/clientset.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	return client
+}
+
+type K8sClient struct {
+	sync.Mutex
+	inited bool
+
+	ClientSet *kubernetes.Clientset
+
+	Ep   EpClient
```
Comment (mxiamxia) [inline internal/aws/k8scommon/k8sclient/clientset.go:38]: What is this client? Let's try to fix the comments on the code. It will be easier for the upstream PR review.

Comment (pxaws) [reply internal/aws/k8scommon/k8sclient/clientset.go:38]: Sure. I will add comments gradually.

**Comment by anuraaga** [comment]
Can you describe what the new receiver specifically does at a high level? (not just product name)

Also, what is your strategy with the change? I skimmed through and saw a lot of k8s-related code which I think the collector should already have, we can't add duplicate functionality.

**Comment by pxaws** [comment]
> Can you describe what the new receiver specifically does at a high level? (not just product name)
> 

Sure. This PR is actually working in progress. I will add more description later. 


> Also, what is your strategy with the change? I skimmed through and saw a lot of k8s-related code which I think the collector should already have, we can't add duplicate functionality.

Yes, there is some upstream k8s-related code scattered in multiple places like `k8sclusterreceiver`, `kubeletstatsreceiver`, and `k8sprocessor`. But I feel like none of them is generic enough so that we can use them directly. That's why I want to migrate the relevant code from cloudwatch agent: https://github.com/aws/amazon-cloudwatch-agent/tree/master/internal/k8sCommon. For me, the quickest way to support container insight use case is to reuse those code. However in the long run, extracting the common k8s utils out from multiple receiver and processors is definitely the right way to go ahead. I think we can help the community to do refactoring in future. But this would be out of scope for current work.

File: internal/receiver/awscontainerinsightreceiver/factory.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			NameVal: typeStr,
+		},
+		CollectionInterval:    defaultCollectionInterval,
+		ContainerOrchestrator: defaultContainerOrchestrator,
+	}
+}
+
+// CreateMetricsReceiver creates an AWS Container Insight receiver.
+func createMetricsReceiver(
+	ctx context.Context,
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/factory.go:47]: replace `ctx` to `_`

File: internal/receiver/awscontainerinsightreceiver/receiver.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// collectData collects container stats from Amazon ECS Task Metadata Endpoint
+func (acir *awsContainerInsightReceiver) collectData(ctx context.Context) error {
+	var mds []pdata.Metrics
+	if acir.cadvisor != nil {
+		mds = append(mds, acir.cadvisor.GetMetrics()...)
+	}
+
+	if acir.k8sapiserver != nil {
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/receiver.go:99]: if `cadvisor` or `k8sapiserver` is nil, should we fail the receiver?

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	return k
+}
+
+func (k *K8sAPIServer) GetMetrics() []pdata.Metrics {
+	var result []pdata.Metrics
+	if k.leading {
+		log.Printf("D! collect data from K8s API Server...")
+		timestamp := strconv.FormatInt(time.Now().UnixNano(), 10)
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:54]: in OTel, we use `timestampMs` as variable name for timestamp in MS.  :)

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		client := k8sclient.Get()
+
+		fields := map[string]interface{}{
+			"cluster_failed_node_count": client.Node.ClusterFailedNodeCount(),
+			"cluster_node_count":        client.Node.ClusterNodeCount(),
+		}
+		tags := map[string]string{
+			common.MetricType: common.TypeCluster,
+			common.Timestamp:  timestamp,
+			common.Version:    "0",
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:64]: do we still need this?

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	if k.leading {
+		log.Printf("D! collect data from K8s API Server...")
+		timestamp := strconv.FormatInt(time.Now().UnixNano(), 10)
+		client := k8sclient.Get()
+
+		fields := map[string]interface{}{
+			"cluster_failed_node_count": client.Node.ClusterFailedNodeCount(),
+			"cluster_node_count":        client.Node.ClusterNodeCount(),
+		}
+		tags := map[string]string{
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:61]: do you think if we use `attributes` instead of `tags` making better sense in OTel?

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	return result
+}
+
+func (k *K8sAPIServer) start() error {
+	var ctx context.Context
+	ctx, k.cancel = context.WithCancel(context.Background())
+
+	k.nodeName = os.Getenv("HOST_NAME")
+	if k.nodeName == "" {
+		return errors.New("Missing environment variable HOST_NAME. Please check your YAML config.")
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:116]: the golang error msg starts with lowercase. :)

File: internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+		// Override the default cAdvisor housekeeping interval.
+		"housekeeping_interval": defaultHousekeepingInterval.String(),
+		// override other defaults (in future)
+	}
+	for name, defaultValue := range flagOverrides {
+		if f := flag.Lookup(name); f != nil {
+			f.DefValue = defaultValue
+			f.Value.Set(defaultValue)
+		} else {
+			logger.Warn("Expected cAdvisor flag not found", zap.String("name", name))
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go:62]: we dont' set `housekeeping_interval` in OTEL collector args. will we always have the warning msg here?

File: internal/receiver/awscontainerinsightreceiver/host/machineinfo.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	if m.instanceId == "" {
+		m.instanceId = m.getAwsMetadata("instance-id")
+	}
+
+	if m.instanceType == "" {
+		m.instanceType = m.getAwsMetadata("instance-type")
+	}
+}
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/host/machineinfo.go:69]: it looks these two can be merged to one call

File: internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+func (c *Cadvisor) GetMetrics() []pdata.Metrics {
+	c.logger.Debug("collect data from cadvisor...")
+	var result []pdata.Metrics
+	var containerinfos []*cinfo.ContainerInfo
+	var err error
+
+	if c.manager == nil && c.initManager() != nil {
+		panic("Cannot initiate manager")
+	}
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go:157]: Q: do we need to call `c.initManager()` for every refresh? can we move it into `receiver.Start()`?

File: internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+	c.logger.Debug("cadvisor containers stats", zap.Int("size", len(containerinfos)))
+	results := processContainers(containerinfos, c.machineInfo, c.containerOrchestrator)
+
+	c.decorateMetrics(results)
+
+	for _, cadvisorMetric := range results {
+		rms := common.ConvertToOTLPMetrics(cadvisorMetric.GetFields(), cadvisorMetric.GetTags(), c.logger)
+		result = append(result, pdata.Metrics(rms))
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go:176]: change to `result = append(result, rms)`?

File: internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+	var result []pdata.Metrics
+	var containerinfos []*cinfo.ContainerInfo
+	var err error
+
+	if c.manager == nil && c.initManager() != nil {
+		panic("Cannot initiate manager")
+	}
+
+	req := &cinfo.ContainerInfoRequest{
+		NumStats: 1,
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go:160]: Can I understand why only `NumStats: 1`?

File: internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+	tags := map[string]string{}
+
+	var containerType string
+	if info.Name != "/" {
+		// Only a container has all these three labels set.
+		containerName := info.Spec.Labels[containerNameLabel]
+		namespace := info.Spec.Labels[namespaceLable]
+		podName := info.Spec.Labels[podNameLable]
+		podId := info.Spec.Labels[podIdLable]
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go:96]: typos: `Lable`

File: internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+		if containerName == "" || namespace == "" || podName == "" {
+			return result, pKey
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go:98]: add a `warn` log?

File: internal/receiver/awscontainerinsightreceiver/cadvisor/extractors/extractor.go [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+type MetricExtractor interface {
+	HasValue(*cinfo.ContainerInfo) bool
+	GetValue(*cinfo.ContainerInfo, MachineInfoProvider, string) []*CAdvisorMetric
+}
+
+type CAdvisorMetric struct {
+	fields map[string]interface{}
+	tags   map[string]string
```
Comment (mxiamxia) [inline internal/receiver/awscontainerinsightreceiver/cadvisor/extractors/extractor.go:40]: Pls help to add comments how we define these 2 variables?

File: internal/processor/awscontainerinsightprocessor/processor.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/aws-observability/aws-otel-collector/internal/processor/awscontainerinsightprocessor/stores"
+	"github.com/aws-observability/aws-otel-collector/internal/processor/awscontainerinsightprocessor/structuredlogsadapter"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+)
+
+const (
+	LogGroupNameTag  = "log_group_name"
+	LogStreamNameTag = "log_stream_name"
+)
```
Comment (mxiamxia) [inline internal/processor/awscontainerinsightprocessor/processor.go:19]: remove if not used?

File: internal/processor/awscontainerinsightprocessor/processor.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+type awscontainerinsightprocessor struct {
+	logger          *zap.Logger
+	started         bool
+	stores          []stores.K8sStore
+	shutdownC       chan bool
+	TagService      bool   `toml:"tag_service"`
+	ClusterName     string `toml:"cluster_name"`
+	HostIP          string `toml:"host_ip"`
+	NodeName        string `toml:"node_name"`
+	PrefFullPodName bool   `toml:"prefer_full_pod_name"`
```
Comment (mxiamxia) [inline internal/processor/awscontainerinsightprocessor/processor.go:30]: could you comment why do we need to have `PrefFullPodName` here?

File: internal/processor/awscontainerinsightprocessor/stores/sampleMountFile/mounts [hunk 19]
```diff
@@ -0,0 +1,2 @@
+/dev/nvme1n1 /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/aws/us-west-2b/vol-0d9f0816149eb2050 ext4 rw,relatime,data=ordered 0 0
+/dev/nvme1n1 /var/lib/kubelet/pods/df570351-2e4c-11e9-95ea-0a695d7ce286/volumes/kubernetes.io~aws-ebs/pvc-df563cf6-2e4c-11e9-95ea-0a695d7ce286 ext4 rw,relatime,data=ordered 0 0
```
Comment (mxiamxia) [inline internal/processor/awscontainerinsightprocessor/stores/sampleMountFile/mounts:2]: is this file still needed? is it just for testing?

File: internal/receiver/awscontainerinsightreceiver/receiver.go (hunk 4 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/receiver.go:99]: I think we want the receiver to be robust and continue to run if only one of the component fails. However when both are failed, we should fail the receiver. I will add code to do that.

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go (hunk 5 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:54]: OK. I will change to use `timestampNs`

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go (hunk 6 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:64]: The emf logs from cloudwatch agent contain "Version: 0". We need to make them the same for otel.

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go (hunk 7 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:61]: Sure. Will change it.

File: internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go (hunk 8 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go:116]: `HOST_NAME` is actually defined in the yaml file for cloudwatch agent

https://github.com/aws-samples/amazon-cloudwatch-container-insights/blob/master/k8s-deployment-manifest-templates/deployment-mode/daemonset/container-insights-monitoring/cwagent/cwagent-daemonset.yaml#L36-L39

File: internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go (hunk 9 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go:62]: This code is used to override some default setting for cadvisor because it doesn't provide an api. Some flag might not be there with new version of cadvisor (if we later upgrade cadvisor). So it makes sense to keep the check here.

File: internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go (hunk 12 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go:176]: Right! The return value is already of type `pdata.Metrics`

File: internal/receiver/awscontainerinsightreceiver/host/machineinfo.go (hunk 10 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/host/machineinfo.go:69]: Right! This will reduce the api calls.

File: internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go (hunk 14 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go:96]: Thanks! I will change it

File: internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go (hunk 15 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/cadvisor/container_info_processor.go:98]: Sure.

File: internal/processor/awscontainerinsightprocessor/processor.go (hunk 17 above)
Comment (pxaws) [reply internal/processor/awscontainerinsightprocessor/processor.go:19]: yes.

File: internal/receiver/awscontainerinsightreceiver/cadvisor/extractors/extractor.go (hunk 16 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/cadvisor/extractors/extractor.go:40]: sure

File: internal/processor/awscontainerinsightprocessor/processor.go (hunk 18 above)
Comment (pxaws) [reply internal/processor/awscontainerinsightprocessor/processor.go:30]: This is an option provided by cloudwatch agent. If it is true, the full pod name (with the hash) will be printed out. See details here: https://github.com/aws/amazon-cloudwatch-agent/blob/master/plugins/processors/k8sdecorator/stores/podstore_test.go#L537-L556

File: internal/processor/awscontainerinsightprocessor/stores/sampleMountFile/mounts (hunk 19 above)
Comment (pxaws) [reply internal/processor/awscontainerinsightprocessor/stores/sampleMountFile/mounts:2]: It's for testing only. I will remove it for now. When we send PR to upstream repo `opentelemetry-collector-contrib`, we will need to add more unit tests and might need this file.

File: internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go (hunk 13 above)
Comment (pxaws) [reply internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go:160]: This is used to control the number of stats objects cadvisor api returns. Within each collection interval of cadvisor, multiple stats will be generated for each container. However we don't need stats with that fine granularity, so we only retrieve one of them. See details here: https://github.com/google/cadvisor/blob/master/info/v1/container.go#L109-L111

**Comment by mxiamxia** [comment]
> Can you describe what the new receiver specifically does at a high level? (not just product name)

+1, pls help to create README for each components and describe the design and usage in high level.

> 
> Also, what is your strategy with the change? I skimmed through and saw a lot of k8s-related code which I think the collector should already have, we can't add duplicate functionality.

This is one big piece of code the team ported from CWAgent for supporting ContainerInsights. We decided to port the code and host it in ADOT collector in temporary and ship it for preview in April. In Q2, we will split these components to smaller PRs and send to upstream for review and merge. (should dedup the code on the existing components if there is any)

**Review by mxiamxia — COMMENTED** [review]
Generally, LGTM from the implementation. But we'll need to improve the code doc and README. Please create an issue before merging these component for TODOs

**Review by mxiamxia — APPROVED** [review]
Thanks!
